milli/src/update/word_prefix_pair_proximity_docids.rs

```diff
@@ -169,89 +368,348 @@ impl<'t, 'u, 'i> WordPrefixPairProximityDocids<'t, 'u, 'i> {
+        // and if the current word2 starts with the same letter, then there is also no potential
+        // prefixes for the current word2, and we can skip to the next iteration
+        if empty_prefixes && !word2_start_different_than_prev {
+            continue;
+        }
+
+        // if word1 is different than the previous word1 OR if the start of word2 is different
+        // than the previous start of word2, then we'll need to flush the batch
+        let word1_different_than_prev = word1 != batch.word1;
+        if word1_different_than_prev || word2_start_different_than_prev {
+            batch.flush(&mut merge_buffer, &mut insert)?;
+            // don't forget to reset the value of batch.word1 and prev_word2_start
+            if word1_different_than_prev {
+                prefix_search_start.0 = 0;
+                batch.word1.clear();
+                batch.word1.extend_from_slice(word1);
+            }
+            if word2_start_different_than_prev {
+                // word2_start_different_than_prev == true
+                prev_word2_start = word2[0];
+            }
+            // Optimisation: find the search start in the prefix trie to iterate over the prefixes of word2
+            empty_prefixes = !prefixes.set_search_start(word2, &mut prefix_search_start);
+        }
+
+        if !empty_prefixes {
+            // All conditions are satisfied, we can now insert each new prefix of word2 into the batch
+            prefixes.for_each_prefix_of(
+                word2,
+                &mut prefix_buffer,
+                &prefix_search_start,
+                |prefix_buffer| {
+                    let prefix_len = prefix_buffer.len();
+                    prefix_buffer.push(0);
+                    prefix_buffer.push(proximity);
+                    batch.insert(&prefix_buffer, data.to_vec());
+                    prefix_buffer.truncate(prefix_len);
+                },
+            );
+            prefix_buffer.clear();
+        }
+    }
+    batch.flush(&mut merge_buffer, &mut insert)?;
+    Ok(())
+}
+/**
+A map structure whose keys are (prefix, proximity) and whose values are vectors of bitstrings (serialized roaring bitmaps).
+The keys are sorted and conflicts are resolved by merging the vectors of bitstrings together.
+
+It is used to ensure that all ((word1, prefix, proximity), docids) are inserted into the database in sorted order and efficiently.
+
+The batch is flushed as often as possible, when we are sure that every (word1, prefix, proximity) key derived from its content
+can be inserted into the database in sorted order. When it is flushed, it calls a user-provided closure with the following arguments:
+- key   : (word1, prefix, proximity) as bytes
+- value : merged roaring bitmaps from all values associated with (prefix, proximity) in the batch, serialised to bytes
+*/
+#[derive(Default)]
+struct PrefixAndProximityBatch {
+    word1: Vec<u8>,
+    batch: Vec<(Vec<u8>, Vec<Cow<'static, [u8]>>)>,
+}
```

Comment:
I am thinking about a potential issue with this data structure: could you prove to me that it doesn't consume to much memory? I mean, if you tell me that the tree we not grow much as it is flush every time the word1 changes it is OK.

Reply:
It is flushed every time the word1 changes and every time the first letter of word2 changes, so it never gets big.